api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java

```diff
@@ -387,18 +406,181 @@ public boolean supportsPropertyName(String gpName) {
+			}
+			savedPatient = persistImportedMpiPatient(foundPatient);
+		} catch (RuntimeException e) {
+			servePdqExceptionAndThrowAgain(e,
+					prepareParameters(patientIdentifier, patientIdentifierTypeUuid),
+					PdqErrorHandlingService.PERSIST_MPI_PATIENT_DESTINATION);
+		}
+		return savedPatient;
+	}
+
+	private void servePdqExceptionAndThrowAgain(RuntimeException e, String message,
+			String destination) {
+		ErrorHandlingService errorHandler = registrationCoreProperties.getPdqErrorHandlingService();
+		if (errorHandler == null) {
+			throw new MpiException(message + " with not configured PDQ error handler", e);
 		} else {
-			//should not pass here since "importPatient" performs only when MpiProvider is not null
-			throw new APIException("Should not perform 'fetchMpiPatient' when MPI is disabled");
+			errorHandler.handle(message, destination,true,
+					ExceptionUtils.getFullStackTrace(e));
+			throw new MpiException(message, e);
+		}
+	}
+
+	private String prepareParameters(String identifier, String identifierTypeUuid) {
+		FetchingMpiPatientParameters parameters = new FetchingMpiPatientParameters(identifier,
+				identifierTypeUuid);
+		try {
+			return new ObjectMapper().writeValueAsString(parameters);
+		} catch (IOException e) {
+			throw new RuntimeException("Cannot prepare parameters for OutgoingMessageException", e);
+		}
+	}
+
+	@Override
+	public Patient findByPatientIdentifier(String patientIdentifier, String patientIdentifierTypeUuid) {
+		PatientIdentifierType idType = patientService.getPatientIdentifierTypeByUuid(patientIdentifierTypeUuid);
+		if (idType == null) {
+			throw new APIException(String.format("PatientIdentifierType is missing: UUID %s",
+					patientIdentifierTypeUuid));
+		} else {
+			//Currently method getPatients() does not take into consideration the identifierTypes,
+			//so it needs to be filtered anyway
+			List<Patient> patients = new ArrayList<Patient>();
+			for (Patient p : patientService.getPatients(null, patientIdentifier,
+					Collections.singletonList(idType), true)) {
+				for (PatientIdentifier pi : p.getIdentifiers()) {
+					if (pi.getIdentifierType().equals(idType)) {
+						patients.add(p);
+						break;
+					}
+				}
+			}
+
+			if (CollectionUtils.isEmpty(patients)) {
+				return null;
+			} else {
+				Patient patient = patients.get(0);
+				dao.getSessionFactory().getCurrentSession().refresh(patient);
+				return patient;
+			}
+			//TODO Add error if more than one patient is has that identifier!
 		}
 	}
 
+	private Patient persistImportedMpiPatient(MpiPatient mpiPatient) {
+		// TODO: Fix bug MPI-11 Updating of Locally generated ID in MPI after MPI import not occurring
+		String openMrsIdTypeUuid = adminService.getGlobalProperty(RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_UUID);
+
+		PatientIdentifierType openMrsIdType = patientService.getPatientIdentifierTypeByUuid(openMrsIdTypeUuid);
+		Patient patient;
+		if (mpiPatient.getPatientIdentifier(openMrsIdType) == null) {
```

Comment:
Doesn't Patient.AddIdentifier() automatically do this anyways?

Reply:
is the "this" you are referring to: checking if the patient already has an OpenMRS ID as specified by the GP and then generating one if there isn't? because I don't think it does that.

if you are saying that just that if statement is unnecessary and to just replace the lines 511 - 514 with:
`PatientIdentifierType openMrsIdType = patientService.getPatientIdentifierTypeByUuid(openMrsIdTypeUuid);
		PatientIdentifier localId = validateOrGenerateIdentifier(mpiPatient.getPatientIdentifier(openMrsIdType).getIdentifier(), mpiPatient.getPatientIdentifier(openMrsIdType).getLocation());
		mpiPatient.addIdentifier(localId);`

then that should work since addIdentifier will just ignore the identifier if it already exists so validateOrGenerateIdentifier will just do a validation. Ok I have changed it to this as it is slightly cleaner, will test to make sure the behaviour is expected

Reply:
Actually I can't do this because if mpiPatient.getPatientIdentifier(openMrsIdType) is null, calling getIdentifier() or getLocation() on it will lead to a null pointer exception